src/go/rpk/pkg/config/weak.go

```diff
@@ -0,0 +1,592 @@
+}
+
+// socketAddresses is an intermediary one_or_many type to be used
+// during our transition to strictly typed configuration parameters.
+// This type will:
+//   - parse an array of SocketAddress
+//   - parse a single SocketAddress to an array.
+type socketAddresses []SocketAddress
+
+func (s *socketAddresses) UnmarshalYAML(n *yaml.Node) error {
+	var multi []SocketAddress
+	err := n.Decode(&multi)
+	if err == nil {
+		*s = multi
+		return nil
+	}
+
+	var single SocketAddress
+	err = n.Decode(&single)
+	if err != nil {
+		return err
+	}
+	fmt.Println("deprecation warning: single element socket addresses will be deprecated in v23.2.1")
+	*s = []SocketAddress{single}
+	return nil
+}
+
+// namedSocketAddresses is an intermediary one_or_many type to be used
+// during our transition to strictly typed configuration parameters.
+// This type will:
+//   - parse an array of NamedSocketAddress
+//   - parse a single NamedSocketAddress to an array.
+type namedSocketAddresses []NamedSocketAddress
+
+func (nsa *namedSocketAddresses) UnmarshalYAML(n *yaml.Node) error {
+	var multi []NamedSocketAddress
+	err := n.Decode(&multi)
+	if err == nil {
+		*nsa = multi
+		return nil
+	}
+
+	var single NamedSocketAddress
+	err = n.Decode(&single)
+	if err != nil {
+		return err
+	}
+	fmt.Println("deprecation warning: single element named socket addresses will be deprecated in v23.2.1")
+	*nsa = []NamedSocketAddress{single}
+	return nil
+}
+
+// serverTLSArray is an intermediary one_or_many type to be used during our
+// transition to strictly typed configuration parameters. This type will:
+//   - parse an array of ServerTLS
+//   - parse a single ServerTLS to an array.
+type serverTLSArray []ServerTLS
+
+func (s *serverTLSArray) UnmarshalYAML(n *yaml.Node) error {
+	var multi []ServerTLS
+	err := n.Decode(&multi)
+	if err == nil {
+		*s = multi
+		return nil
+	}
+
+	var single ServerTLS
+	err = n.Decode(&single)
+	if err != nil {
+		return err
+	}
+	fmt.Println("deprecation warning: single element server TLS will be deprecated in v23.2.1")
+	*s = []ServerTLS{single}
+	return nil
+}
+
+// seedServers is an intermediary one_or_many type to be used during our
+// transition to strictly typed configuration parameters. This type will:
+//   - parse an array of SeedServer
+//   - parse a single SeedServer to an array.
+type seedServers []SeedServer
+
+func (ss *seedServers) UnmarshalYAML(n *yaml.Node) error {
+	var multi []SeedServer
+	err := n.Decode(&multi)
+	if err == nil {
+		*ss = multi
+		return nil
+	}
+
+	var single SeedServer
+	err = n.Decode(&single)
+	if err != nil {
+		return err
+	}
+	fmt.Println("deprecation warning: single element seed server will be deprecated in v23.2.1")
+	*ss = []SeedServer{single}
+	return nil
+}
+
+// Custom unmarshallers for all the config related types.
+
+func (c *Config) UnmarshalYAML(n *yaml.Node) error {
+	var internal struct {
+		file *Config
+
+		NodeUUID             weakString      `yaml:"node_uuid,omitempty"`
```

Comment:
Minor nit, but I don't think we need `,omitempty` in this custom unmarshal -- omitempty is only relevant when marshaling.